File: DuplicationInArray/FindDuplication.py
def duplicate(numbers, duplication):
    if not numbers or len(numbers) <= 0:
        return False

    for i in numbers:
        if i < 0 or i >= len(numbers):
            return False

    for i in range(len(numbers)):
        while numbers[i] != i:
            if numbers[i] == numbers[numbers[i]]:
                duplication[0] = numbers[i]
                return True
            else:
                # numbers[i], numbers[numbers[i]] = numbers[numbers[i]], numbers[i]
                temp = numbers[i]
                numbers[i], numbers[temp] = numbers[temp], numbers[i]
        return False


# L = [2, 3, 5, 4, 3, 2, 6, 7]
# LL = [0]
# duplicate(L, LL)
# print(LL)

File: 03_01_DuplicationInArray/FindDuplication.cpp
class Solution
{
  public:
    // Parameters:
    //        numbers:     an array of integers
    //        length:      the length of array numbers
    //        duplication: (Output) the duplicated number in the array number
    // Return value:       true if the input is valid, and there are some duplications in the array number
    //                     otherwise false
    bool duplicate(int numbers[], int length, int *duplication)
    {
        if (numbers == nullptr || length <= 0)
        {
            return false;
        }
        for (int i = 0; i < length; i++)
        {
            if (numbers[i] >= length || numbers[i] < 0)
            {
                return false;
            }
        }
        for (int i = 0; i < length; i++)
        {
            while (numbers[i] != i)
            {
                if (numbers[i] == numbers[numbers[i]])
                {
                    *duplication = numbers[i];
                    return true;
                }
                else
                {
                    // #include <algorithm>
                    // using namespace std;
                    swap(numbers[i], numbers[numbers[i]]);
                }
            }
        }
        return false;
    }
};
